src_c/geometry.c

```diff
@@ -8,6 +8,106 @@
 
+static PyObject *
+pg_raycast(PyObject *_null, PyObject *const *args, Py_ssize_t nargs)
+{
+    PyObject **farr;
+    Py_ssize_t farr_length;
+    Py_ssize_t loop;
+    double max_t;
+    pgLineBase line;
+
+    if (nargs == 4) {
+        if (!pg_TwoDoublesFromObj(args[0], &line.x1, &line.y1)) {
+            return RAISE(PyExc_TypeError,
+                         "the origin requires a pair of floats");
+        }
+        double max_dist;
+        if (!pg_DoubleFromObj(args[2], &max_dist)) {
+            return RAISE(PyExc_TypeError, "invalid max distance");
+        }
+        if (max_dist < 0) {
+            max_dist = DBL_MAX;
+        }
+        if (!pg_TwoDoublesFromObj(args[1], &line.x2, &line.y2)) {
+            double angle;
+            if (!pg_DoubleFromObj(args[1], &angle)) {
+                return RAISE(PyExc_TypeError,
+                             "expected a pair of floats or a single float");
+            }
+
+            if (max_dist != DBL_MAX) {
+                line.x2 = line.x1 - cos(angle * PI / 180) * max_dist;
+                line.y2 = line.y1 - sin(angle * PI / 180) * max_dist;
+            }
+            else {
+                line.x2 = line.x1 - cos(angle * PI / 180);
+                line.y2 = line.y1 - sin(angle * PI / 180);
+            }
+        }
+        if (!PySequence_FAST_CHECK(args[3])) {
+            return RAISE(PyExc_TypeError,
+                         "colliders parameter must be a sequence");
+        }
+        farr = PySequence_Fast_ITEMS(args[3]);
+        farr_length = PySequence_Fast_GET_SIZE(args[3]);
+
+        max_t = max_dist / pgLine_Length(&line);
+    }
+    else if (nargs == 2) {
+        if (!pgLine_FromObject(args[0], &line)) {
+            return RAISE(PyExc_TypeError,
+                         "line parameter must be a Line or a LineLike object");
+        }
+
+        max_t = 1;
+
+        farr = PySequence_Fast_ITEMS(args[1]);
+        farr_length = PySequence_Fast_GET_SIZE(args[1]);
+    }
+    else {
+        return RAISE(PyExc_TypeError, "invalid number of arguments");
+    }
+
+    // find the best t
+    double record_t = max_t;
+    double temp_t = 0;
+
+    for (loop = 0; loop < farr_length; loop++) {
+        if (pgCircle_Check(farr[loop])) {
+            if (pgRaycast_LineCircle(&line, &pgCircle_AsCircle(farr[loop]),
+                                     max_t, &temp_t)) {
+                record_t = MIN(record_t, temp_t);
+            }
+        }
+        else if (pgLine_Check(farr[loop])) {
+            if (pgRaycast_LineLine(&line, &pgLine_AsLine(farr[loop]), max_t,
+                                   &temp_t)) {
+                record_t = MIN(record_t, temp_t);
+            }
+        }
+        else if (pgRect_Check(farr[loop])) {
+            if (pgRaycast_LineRect(&line, &pgRect_AsRect(farr[loop]), max_t,
+                                   &temp_t)) {
+                record_t = MIN(record_t, temp_t);
+            }
+        }
```

Comment:
Not that it matters much, though it would shorten the code length a bit, you could store the `farr[loop]` into a `PyObject* obj`
```suggestion
        PyObject *obj = farr[loop];
        if (pgCircle_Check(obj)) {
            if (pgRaycast_LineCircle(&line, &pgCircle_AsCircle(obj),
                                     max_t, &temp_t)) {
                record_t = MIN(record_t, temp_t);
            }
        }
        else if (pgLine_Check(obj)) {
            if (pgRaycast_LineLine(&line, &pgLine_AsLine(obj), max_t,
                                   &temp_t)) {
                record_t = MIN(record_t, temp_t);
            }
        }
        else if (pgRect_Check(obj)) {
            if (pgRaycast_LineRect(&line, &pgRect_AsRect(obj), max_t,
                                   &temp_t)) {
                record_t = MIN(record_t, temp_t);
            }
        }
```